### DISCUSSION TITLE
Could librashader integration help?

### DISCUSSION THREAD

**Discussion by Quackdoc** [Blinue/Magpie#785] [Ideas]
Librashader is a library compatible with retroarch slang shaders https://github.com/SnowflakePowered/librashader.

I wonder if this could be integrated to help expand magpie's shaders, as well as lessen the efforts needed to port future shaders.

**Comment by Blinue** [comment]
Thank you for letting me know about this cool project! Integrating it doesn’t seem difficult, but there are some factors to consider:

1. RetroArch shaders use 3D pipeline, while Magpie shaders use compute pipeline. Switching pipelines is a time-consuming operation.
2. The dll file of librashader is over 17MB in size, and there is no clean way to import it.
3. This would increase the maintenance cost, as we would need to support two types of shaders. I would avoid doing this unless there is a compelling reason.

I think the appropriate way is to develop some form of **plugin** mechanism, and let the user choose whether to integrate this feature. (By the way, [the support for ONNX](https://github.com/Blinue/Magpie/issues/772) could also be made into a plugin.)

I will keep an eye on this project, thank you for your suggestion!

**Comment by odt1** [comment]
Integrating RetroArch shaders would be the be-all and end-all solution to emulating CRT for old PC games. RetroArch shaders like Guest Advanced, CRT-Royale and Mega Bezel Collection are light-years away from anything available from Magpie or even ReShade.

Currently, there is nothing available on PC that fills this specific gap, and Magpie might be just a perfect platform for it. It is uniquely positioned to handle the technical ability to perfectly feed low-res signal into a high-res shader container and it is an absolute end game for CRT emulation and many other use cases that RetroArch state-of-the-art shader library might provide.

Please, please consider adding support for Slang RetroArch shaders and presets, this would reap enormous benefits for a lot of users.

Thank you for your hard work!

**Comment by Quackdoc** [comment]
Since it was brought up, a good point of reference might be https://github.com/hhirtz/mpv-libretro which converts *some* retroarch shaders presets into libplacebo glsl shaders. With a generic glsl shader loading pipeline this could work, or it could be forked to convert shaders into a magpiefx file

**Comment by Blinue** [comment]
Thanks for your interest. I’m currently working on a new renderer with plugin support, which means integrating librashader or other technologies may be possible in the future. I’m hoping to have it completed within the next few months.
